src/resources.js

```diff
@@ -475,22 +530,64 @@ export class Resources {
       let minTop = -1;
       for (let i = 0; i < changeHeightRequests.length; i++) {
         let request = changeHeightRequests[i];
+        let resource = request.resource;
         let box = request.resource.getLayoutBox();
         if (box.height == request.newHeight) {
           // Nothing to do.
           continue;
         }
-        if (box.top >= 0) {
-          minTop = minTop == -1 ? box.top : Math.min(minTop, box.top);
+
+        // Check resize rules. It will either resize element immediately, or
+        // wait until scrolling stops or will call the fallback.
+        let resize = false;
+        if (request.force || !this.visible_) {
+          // 1. An immediate execution requested or the document is hidden.
+          resize = true;
+        } else if (activeElement && (activeElement == resource.element ||
+              resource.element.contains(activeElement))) {
+          // 2. Active elements are immediately resized. The assumption is that
+          //    the resize is triggered by the user action or soon after.
+          resize = true;
+        } else if (box.bottom >= viewportRect.bottom - offset) {
+          // 3. Elements under viewport are resized immediately.
+          resize = true;
+        } else if (box.bottom <= viewportRect.top + offset) {
+          // 4. Elements above the viewport can only be resized when scrolling
+          //    has stopped, otherwise defer util next cycle.
+          if (isScrollingStopped) {
+            resize = true;
+          } else {
+            // Defer till next cycle.
+            this.changeHeightRequests_.push(request);
+          }
+        } else if (request.newHeight < box.height) {
+          // 5. Resize immediately if the requested size is smaller.
```

Comment:
RE: #1 - yes, something like this is possible with a horrible UX of course. Not sure though - we don't have a lot of options. We can prohibit reducing size for now...

#2: So, no resize of any kind above the current scroll position is allowed while the user is still scrolling. With or without the scroll adjustment I'm intending to keep that. The cost of it is very low and benefit is high.